src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3821,6 +3820,279 @@ internal new NamedTypeSymbol CreateNativeIntegerTypeSymbol(bool signed)
+                => type.TypeKind is TypeKind.Dynamic or TypeKind.Error;
+
+            void checkOperatorKinds()
+            {
+                // Dynamic built-in operators allow virtually all operations with all types.  So we do no further checking here.
+                if (isDynamicOrError(csharpReturnType) ||
+                    isDynamicOrError(csharpLeftType) ||
+                    isDynamicOrError(csharpRightType))
+                {
+                    return;
+                }
+
+                // Use fast-path check to see if this types are ok.
+                var binaryKind = OperatorFacts.SyntaxKindToBinaryOperatorKind(SyntaxFacts.GetBinaryExpression(syntaxKind));
+                var easyOutBinaryKind = OverloadResolution.BinopEasyOut.OpKind(binaryKind, csharpLeftType, csharpRightType);
+
+                if (easyOutBinaryKind != BinaryOperatorKind.Error)
+                {
+                    var signature = this.builtInOperators.GetSignature(easyOutBinaryKind);
+                    if (TypeSymbol.Equals(csharpReturnType, signature.ReturnType, TypeCompareKind.ConsiderEverything) &&
+                        TypeSymbol.Equals(csharpLeftType, signature.LeftType, TypeCompareKind.ConsiderEverything) &&
+                        TypeSymbol.Equals(csharpRightType, signature.RightType, TypeCompareKind.ConsiderEverything))
+                    {
+                        return;
+                    }
+                }
+
+                // bool operator ==(object, object) is legal.
+                // bool operator !=(object, object) is legal.
+                // bool operator ==(Delegate, Delegate) is legal.
+                // bool operator !=(Delegate, Delegate) is legal.
+                if (binaryKind is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual &&
+                    csharpReturnType.SpecialType is SpecialType.System_Boolean)
+                {
+                    if ((csharpLeftType.SpecialType is SpecialType.System_Object && csharpRightType.SpecialType is SpecialType.System_Object) ||
+                        (csharpLeftType.SpecialType is SpecialType.System_Delegate && csharpRightType.SpecialType is SpecialType.System_Delegate))
+                    {
+                        return;
+                    }
+                }
+
+                // Actual delegates have several operators that can be used on them.
+                if (csharpLeftType.TypeKind is TypeKind.Delegate &&
+                    TypeSymbol.Equals(csharpLeftType, csharpRightType, TypeCompareKind.ConsiderEverything))
+                {
+                    // bool operator ==(SomeDelegate, SomeDelegate) is legal.
+                    // bool operator !=(SomeDelegate, SomeDelegate) is legal.
+                    if (binaryKind is BinaryOperatorKind.Equal or BinaryOperatorKind.NotEqual &&
+                        csharpReturnType.SpecialType == SpecialType.System_Boolean)
+                    {
+                        return;
```

Comment:
>return [](http://example.com/codeflow?start=24&length=6)

Is this code path reachable? It looks like comparison of delegates is handled above. Together with object comparison.

Reply:
one is literally System.Delegate, teh other are acutal delegates (like Action).